File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/NCameraUpdate.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map

import com.naver.maps.geometry.LatLng
import com.naver.maps.geometry.LatLngBounds
import com.naver.maps.map.CameraAnimation
import com.naver.maps.map.CameraUpdate
import com.naver.maps.map.CameraUpdateParams
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asLong
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asCameraAnimation
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLngBounds
import dev.note11.flutter_naver_map.flutter_naver_map.model.base.NEdgeInsets
import dev.note11.flutter_naver_map.flutter_naver_map.model.base.NPoint

internal data class NCameraUpdate(
    val signature: String,
    val target: LatLng? = null,
    val zoom: Double? = null,
    val zoomBy: Double? = null,
    val tilt: Double? = null,
    val tiltBy: Double? = null,
    val bearing: Double? = null,
    val bearingBy: Double? = null,
    val bounds: LatLngBounds? = null,
    val boundsPadding: NEdgeInsets? = null,
    val pivot: NPoint? = null, // change to NPoint
    val animation: CameraAnimation,
    val duration: Long,
    val reason: Int? = null,
) {

    fun toCameraUpdate(): CameraUpdate {
        val cameraUpdate = when (signature) {
            "withParams" -> CameraUpdate.withParams(CameraUpdateParams().apply {
                if (target != null) scrollTo(target)
                if (zoom != null) zoomTo(zoom)
                if (zoomBy != null) zoomBy(zoomBy)
                if (tilt != null) tiltTo(tilt)
                if (tiltBy != null) tiltBy(tiltBy)
                if (bearing != null) rotateTo(bearing)
                if (bearingBy != null) rotateBy(bearingBy)
            })
            "fitBounds" -> boundsPadding?.run {
                CameraUpdate.fitBounds(bounds!!, leftPx, topPx, rightPx, bottomPx)
            } ?: CameraUpdate.fitBounds(bounds!!)
            else -> throw IllegalArgumentException("Unknown signature: $signature")
        }

        return cameraUpdate.animate(animation, duration)
            .run { pivot?.let { pivot(it.toPointF()) } ?: this }
            .run { reason?.let { reason(it) } ?: this }
    }

    companion object {
        fun fromMessageable(args: Any): NCameraUpdate = args.asMap().let { map ->
            return NCameraUpdate(
                signature = map["sign"].toString(),
                target = map["target"]?.asLatLng(),
                zoom = map["zoom"]?.asDouble(),
                zoomBy = map["zoomBy"]?.asDouble(),
                tilt = map["tilt"]?.asDouble(),
                tiltBy = map["tiltBy"]?.asDouble(),
                bearing = map["bearing"]?.asDouble(),
                bearingBy = map["bearingBy"]?.asDouble(),
                bounds = map["bounds"]?.asLatLngBounds(),
                boundsPadding = map["boundsPadding"]?.let(NEdgeInsets::fromMessageable),
                pivot = map["pivot"]?.let(NPoint::fromMessageable),
                animation = map["animation"]!!.asCameraAnimation(),
                duration = map["duration"]!!.asLong(),
                reason = map["reason"]?.asInt()
            )
        }
    }
}

File: ios/Classes/model/map/NCameraUpdate.swift
import NMapsMap

internal struct NCameraUpdate {
    let signature: String
    let target: NMGLatLng?
    let zoom: Double?
    let zoomBy: Double?
    let tilt: Double?
    let tiltBy: Double?
    let bearing: Double?
    let bearingBy: Double?
    let bounds: NMGLatLngBounds?
    let boundsPadding: NEdgeInsets?
    let pivot: NPoint?
    let animation: NMFCameraUpdateAnimation
    let duration: Int // mill seconds. not Second.
    let reason: Int?

    var cameraUpdate: NMFCameraUpdate {
        get throws {
            let cameraUpdate: NMFCameraUpdate

            switch signature {
            case "withParams": cameraUpdate = NMFCameraUpdate(params: params)
            case "fitBounds": cameraUpdate = boundsPadding != nil
                    ? NMFCameraUpdate(fit: bounds!, paddingInsets: boundsPadding!.uiEdgeInsets)
                    : NMFCameraUpdate(fit: bounds!)
            default: throw NSError()
            }

            if let pivot = pivot {
                cameraUpdate.pivot = pivot.cgPoint
            }
            if let reason = reason {
                cameraUpdate.reason = Int32(reason)
            }

            cameraUpdate.animation = animation
            cameraUpdate.animationDuration = (Double(duration) / 1000)

            return cameraUpdate
        }
    }

    private var params: NMFCameraUpdateParams {
        let p = NMFCameraUpdateParams()
        if let target = target {
            p.scroll(to: target)
        }
        if let zoom = zoom {
            p.zoom(to: zoom)
        }
        if let zoomBy = zoomBy {
            p.zoom(by: zoomBy)
        }
        if let tilt = tilt {
            p.tilt(to: tilt)
        }
        if let tiltBy = tiltBy {
            p.tilt(by: tiltBy)
        }
        if let bearing = bearing {
            p.rotate(to: bearing)
        }
        if let bearingBy = bearingBy {
            p.rotate(by: bearingBy)
        }
        return p
    }

    static func fromMessageable(_ v: Any) -> NCameraUpdate {
        let d = asDict(v)
        return NCameraUpdate(
                signature: asString(d["sign"]!),
                target: castOrNull(d["target"], caster: asLatLng),
                zoom: castOrNull(d["zoom"], caster: asDouble),
                zoomBy: castOrNull(d["zoomBy"], caster: asDouble),
                tilt: castOrNull(d["tilt"], caster: asDouble),
                tiltBy: castOrNull(d["tiltBy"], caster: asDouble),
                bearing: castOrNull(d["bearing"], caster: asDouble),
                bearingBy: castOrNull(d["bearingBy"], caster: asDouble),
                bounds: castOrNull(d["bounds"], caster: asLatLngBounds),
                boundsPadding: castOrNull(d["boundsPadding"], caster: NEdgeInsets.fromMessageable),
                pivot: castOrNull(d["pivot"], caster: NPoint.fromMessageable),
                animation: asCameraAnimation(d["animation"]!),
                duration: asInt(d["duration"]!),
                reason: castOrNull(d["reason"], caster: asInt)
        )
    }
}
